### PR TITLE
Mojo::Date: negative & fractional epoch

### PR DESCRIPTION
**Pull request by aferreira**
### Summary
This is a revision of https://github.com/kraih/mojo/pull/1079. 

The test failures pointed in https://github.com/kraih/mojo/pull/1079#issuecomment-291487534 turn out to be a problem with Mojo::Date::to_string() with fractional epoch, and not an effect of accepting negative epoch.

This PR includes fixes to to_string() and to_datetime() in this regard, and brings the change to accept negative epoch again.

### Motivation
Same as https://github.com/kraih/mojo/pull/1079.

Perl documentation (http://perldoc.perl.org/perlport.html#gmtime) mentions that `gmtime` has portability issues because of how floating point numbers are used in the implementation. I think we hit that when [the build](https://travis-ci.org/kraih/mojo/builds/218440119) failed for 5.10.

_Updated (both this description & pull request) after the feedback by @jhthorsen._

The proposed solution is to avoid passing floats to gmtime() – instead
* round the epoch to be used by to_string()
* decompose `$epoch` into `floor($epoch) + $frac` at to_datetime()

The implementation change to to_datetime() gets more complicated as in

```perl
sub to_datetime {

  # RFC 3339 (1994-11-06T08:49:37Z)
  my ($s, $m, $h, $day, $month, $year) = _gmftime(shift->epoch);
  my $str = sprintf '%04d-%02d-%02dT%02d:%02d:%02d', $year + 1900, $month + 1,
    $day, $h, $m, $s;
  return $str . ($s =~ /(\.\d+)$/ ? $1 : '') . 'Z';
}

# gmtime with (possibly) fractional seconds
sub _gmftime {
  return gmtime $_[0] unless $_[0] =~ /^((-?)\d+)(\.\d+)$/;

  # decompose $epoch into $t + $f, where 0 ≤ $f < 1
  my ($t, $f) = $2 ? ($1 - 1, $_[0] - $1 + 1) : ($1, $3);
  my @t = gmtime $t;
  return shift(@t) + $f, @t;
}
```

The code attempts to do as little floating point operations as possible (and only in the case of negative fractional epoch). Among the tests, these illustrate the kind of computation needed to get `to_datetime()` right for this edge case.

```
'0.33'   =>   '1970-01-01T00:00:00.33Z'
'-0.33'  =>  '1969-12-31T23:59:59.67Z'
```

### CODE REVIEW

**Comment by kraih** [comment]
The `to_datetime` code looks quite ugly now, not sure if that's really worth it.

**Comment by kraih** [comment]
Having thought about this some more, i have to vote 👎 this time i'm afraid.

**Comment by aferreira** [comment]
Would this change https://github.com/kraih/mojo/pull/1080/commits/73c0fef4cac0c0e7b5c336f4140f7a9052d2d31d help? I moved the ugliness to an auxiliary function :-/

**Comment by jhthorsen** [comment]
I tried to golf down _gmftime a bit more:

```perl
sub _gmftime {
  return gmtime $_[0] unless $_[0] =~ /^(-?\d+)(\.\d+)$/;
  my ($f, @t) = $_[0] < 0 ? ($_[0] - $1 + 1, gmtime $1 - 1) : ($2, gmtime $1);
  return shift(@t) + $f, @t;
}
```

File: lib/Mojo/Date.pm [hunk 1]
```diff
@@ -54,16 +54,29 @@ sub parse {
...
+  }
+  return gmtime($epoch);
+}
+
+
 sub to_string {
 
   # RFC 7231 (Sun, 06 Nov 1994 08:49:37 GMT)
-  my ($s, $m, $h, $mday, $month, $year, $wday) = gmtime shift->epoch;
+  my ($s, $m, $h, $mday, $month, $year, $wday) = gmtime sprintf('%.0f', shift->epoch);
```
Comment (jhthorsen) [inline lib/Mojo/Date.pm]: What does sprintf do here?

File: lib/Mojo/Date.pm [hunk 2]
```diff
@@ -54,16 +54,29 @@ sub parse {
...
+  my $epoch = shift;
+  if ($epoch =~ /^(-?\d+)(\.\d+)$/) {
+    my $t = ($epoch < 0) ? $1 - 1      : $1;    # floor
+    my $f = ($epoch < 0) ? $epoch - $t : $2;
+    my @t = gmtime($t);
+    $t[0] += $f;
+    return @t;
+  }
+  return gmtime($epoch);
+}
```
Comment (jhthorsen) [inline lib/Mojo/Date.pm]: I've golfed down _gmftime:

```perl
sub _gmftime {
  return &gmtime unless $_[0] =~ /^(-?\d+)(\.\d+)$/;
  my ($f, @t) = $_[0] < 0 ? ($_[0] - $1 + 1, gmtime $1 - 1) : ($2, gmtime $1);
  return shift(@t) + $f, @t;
}
```

File: lib/Mojo/Date.pm (hunk 1 above)
Comment (aferreira) [inline lib/Mojo/Date.pm]: Rounds a float to the nearest integer and passes through an integer unchanged.

File: lib/Mojo/Date.pm (hunk 2 above)
Comment (aferreira) [inline lib/Mojo/Date.pm]: Well played. =) I like the way you compressed the most common case to an efficient expression at the top of the function and the combination of `@t` and `$f` in a single expression.

I am inclined to be a little more verbose to highlight that the important thing here is to decompose `$epoch` into `$t + $f` where `0 ≤ $f < 1`. Something like

```perl
sub _gmftime {
  return &gmtime unless $_[0] =~ /^(-?\d+)(\.\d+)$/;
  # decompose $epoch into $t + $f, where 0 < $f < 1
  my $t = ($_[0] < 0) ? $1 - 1      : $1;
  my $f = ($_[0] < 0) ? $_[0] - $t : $2;
  @t = gmtime $t;
  return shift(@t) + $f, @t;
}
```

Notice this has one less floating point operation (when you compute `$1 -1` twice at `$_[0] - $1 - 1` and `$1 -1`).

Also the variant below can avoid an extra floating point operation above that is caused by the numeric comparison `$_[0] < 0`.

```perl
sub _gmftime {
  return &gmtime unless $_[0] =~ /^((-?)\d+)(\.\d+)$/;
  # decompose $epoch into $t + $f, where 0 < $f < 1
  my $t = $2 ? $1 - 1      : $1;
  my $f = $2 ? $_[0] - $t : $3;
  @t = gmtime $t;
  return shift(@t) + $f, @t;
}
```
Does it look worth?

File: lib/Mojo/Date.pm (hunk 1 above)
Comment (aferreira) [inline lib/Mojo/Date.pm]: Just to make clear, the rounding is important to make sure `gmtime` is called with an integer argument (which is portable). I mentioned that at the start of the "Summary" of this ticket.

Comment (jhthorsen) [inline lib/Mojo/Date.pm]: Thank you 👍

File: lib/Mojo/Date.pm (hunk 2 above)
Comment (aferreira) [inline lib/Mojo/Date.pm]: I think a good compromise would be:

```perl
# gmtime with (possibly) fractional seconds
sub _gmftime {
  return &gmtime unless $_[0] =~ /^((-?)\d+)(\.\d+)$/;
  # decompose $epoch into $t + $f, where 0 ≤ $f < 1
  my ($t, $f) = $2 ? ($1 - 1, $_[0] - $1 + 1) : ($1, $3)
  @t = gmtime $t;
  return shift(@t) + $f, @t;
}
```

which gives up on one extra flop for the "rare" case of negative fractional epoch.

**Comment by kraih** [comment]
Thank you for your effort, but i'm afraid this version of the patch didn't reach the required number of votes.
